src/runner.rs

```diff
@@ -70,14 +87,66 @@ impl<'a, E: Externals> Interpreter<'a, E> {
 				)?;
 		}
 
-		let initial_frame = FunctionContext::new(func.clone());
-
 		let mut call_stack = Vec::new();
+		let initial_frame = FunctionContext::new(func.clone());
 		call_stack.push(initial_frame);
 
-		self.run_interpreter_loop(&mut call_stack)?;
+		let return_type = func.signature().return_type();
+
+		Ok(Interpreter {
+			value_stack,
+			call_stack,
+			return_type,
+			state: InterpreterState::Initialized,
+		})
+	}
+
+	pub fn state(&self) -> &InterpreterState {
+		&self.state
+	}
+
+	pub fn start_execution<'a, E: Externals + 'a>(&mut self, externals: &'a mut E) -> Result<Option<RuntimeValue>, Trap> {
+		// Ensure that the VM has not been executed. This is checked in `FuncInvocation::start_execution`.
+		assert!(self.state == InterpreterState::Initialized);
+
+		self.state = InterpreterState::Started;
+		self.run_interpreter_loop(externals)?;
+
+		let opt_return_value = self.return_type.map(|_vt| {
+			self.value_stack.pop()
+		});
+
+		// Ensure that stack is empty after the execution.
+		assert!(self.value_stack.len() == 0);
```

Comment:
What about this assert?
How is that guaranteed that there are no more values?

Reply:
I'm not sure about this either. This assert is added in https://github.com/paritytech/wasmi/pull/98. @pepyakin Do we have guarantees on this or should we change this to a Trap?

Reply:
sorry my bad. Yeah we do, it’s guaranteed by the validation properties. I should have add a comment